### ISSUE TITLE
raft topology: ban removed, replaced and ignored dead nodes early

### ISSUE THREAD

**Issue by patjed41** [scylladb/scylladb#16108]
When we do a raft-based topology operation, we need to ban all nodes expected to be dead at the beginning of the operation, so that they do not disrupt us in the process. Checking if the node is dead cannot be sufficient because, for example, the dead node can become alive again at any time. More specifically, we need to ban:
- the node being removed and the ignored dead nodes during removenode,
- the node being replaced and the ignored dead nodes during replace,
- the ignored dead nodes during bootstrap if we ever decide to add this feature; currently, there is a TODO in `storage_service::join_node_response_handler`: `TODO: ignore_dead_nodes setting for bootstrap`.

**Comment by patjed41** [comment]
cc @kbr-scylla @gleb-cloudius

**Comment by gleb-cloudius** [comment]
> cc @tgrabiec @gleb-cloudius -- this could unblock stuck tablet migrations.
> 
> Tomek suggested that we ban nodes right at the moment of adding the request to the queue.
> 
> Although, with Gleb's new request management (on mailing list), requests may be cancelled.
> 
> It will be kind of weird because we might add a removenode request, the request bans certain nodes from the cluster permanently, and then it might get cancelled.

Yes, it would be unexpected for part of the operation to take effect while another to fail.

> 
> But I don't see another way. Except maybe having a completely separate API just for banning nodes.

What if the use case? How banning nodes will unstuck tablet migrations?

**Comment by kbr-scylla** [comment]
> What if the use case? How banning nodes will unstuck tablet migrations?

https://docs.google.com/document/d/1ATh5i4EMTJYnWERZ-BDH7lpgqwHe0HE-zxPWycW70So/edit?disco=AAABAl_c85o

**Comment by gleb-cloudius** [comment]
> > What if the use case? How banning nodes will unstuck tablet migrations?
> 
> https://docs.google.com/document/d/1ATh5i4EMTJYnWERZ-BDH7lpgqwHe0HE-zxPWycW70So/edit?disco=AAABAl_c85o

This asks for not only ban nodes on rpc level, but actually mark them as not suitable for tablet migration. This asks for a separate API.

**Comment by kbr-scylla** [comment]
Generally mark nodes as permanently dead which would also ban them.
Banning a node is one-way operation anyway

**Comment by gleb-cloudius** [comment]
> Generally mark nodes as permanently dead which would also ban them. Banning a node is one-way operation anyway

We can get rid of "ignore nodes" parameters. Use this new API to mark nodes that you want to kill and them remove then one by one. But killing a node should be a topology operation. The coordinator needs to make sure all nodes know about dead nodes.

**Comment by bhalevy** [comment]
@kostja, @kbr-scylla: this issue is a prerequisite for #16527 that is targeted for 6.0
Can it be prioritized?

**Comment by tgrabiec** [comment]
topology_state_machine now has get_excluded_nodes(). Tablet side (https://github.com/scylladb/scylladb/issues/16527) will integrate with it. So all tablets need is for get_excluded_nodes() to pick up nodes which are banned permanently. Currently, get_excluded_nodes() pick them up from request params during removenode/replace. This is not enough because if nodes go down during tablet migration, nodetool removenode request will be blocked behind tablet migration and we will deadlock. It would best if we changed nodetool removenode to ban the nodes immediately before it gets scheduled for execution by the topology coordinator. This way we wouldn't need to introduce a new API to do that, and dtests would pass without modifications. If we don't do that from nodetool removenode, dtests would need to ensure that migrations are quiesced before downing the nodes, or tell removenode to ban the nodes immediately via param.

**Comment by gleb-cloudius** [comment]
> It would be best if we changed nodetool removenode to ban the nodes immediately before it gets scheduled for execution by the topology coordinator.

But banning a node should also be a kind of a topology operation (because it needs to contact all of the nodes in a redundant manner). So it will also stuck behind tablet migration unless we allow tables migration and (at least some) topology operations  to run in parallel. This is a chicken and egg problem.

And we need to define what would happen if the banning fails (because it was not able to contact all of the nodes for instance). 

Wouldn't it be better to make tablet migration preemptable? IOW allow topology operations while there are migration tasks that do not stream right now (does it make sense)?

**Comment by gleb-cloudius** [comment]
> > > It would be best if we changed nodetool removenode to ban the nodes immediately before it gets scheduled for execution by the topology coordinator.
> > 
> > 
> > But banning a node should also be a kind of a topology operation (because it needs to contact all of the nodes in a redundant manner). So it will also stuck behind tablet migration unless we allow tables migration and (at least some) topology operations to run in parallel. This is a chicken and egg problem.
> 
> Does it have to be more than a group0 commit which updates `excluded_nodes`?

Of course. This is what this issue is about. Banning a node is not just adding them to `excluded_nodes` but making sure that they cannot talk to other nodes in the cluster any longer. Currently it happens at the end of the topology change when nodes entering left state and the bug suggest that it should happen earlier (not sure I agree with that) and also should happen for `excluded_nodes` as well (which I agree with, though I wanted this banning to be temporary). For that we need to communicate with all nodes to make sure all of them know that a certain node is banned.

> > Wouldn't it be better to make tablet migration preemptable? IOW allow topology operations while there are migration tasks that do not stream right now (does it make sense)?
> 
> I'd rather avoid that, because proving correctness is impossible then without global program analysis to consider all possible interactions of other topology operations with tablet migration.

May be we can allow only certain topology operations while there are ongoing migrations.

**Comment by tgrabiec** [comment]
On Mon, Jan 29, 2024 at 1:40 PM Gleb Natapov ***@***.***>
wrote:

> It would be best if we changed nodetool removenode to ban the nodes
> immediately before it gets scheduled for execution by the topology
> coordinator.
>
> But banning a node should also be a kind of a topology operation (because
> it needs to contact all of the nodes in a redundant manner). So it will
> also stuck behind tablet migration unless we allow tables migration and (at
> least some) topology operations to run in parallel. This is a chicken and
> egg problem.
>
> Does it have to be more than a group0 commit which updates excluded_nodes?
>
> Of course. This is what this issue is about. Banning a node is not just
> adding them to excluded_nodes but making sure that they cannot talk to
> other nodes in the cluster any longer. Currently it happens at the end of
> the topology change when nodes entering left state and the bug suggest that
> it should happen earlier (not sure I agree with that) and also should
> happen for excluded_nodes as well (which I agree with, though I wanted
> this banning to be temporary). For that we need to communicate with all
> nodes to make sure all of them know that a certain node is banned.
>
Can hard banning happen lazily as each node learns about excluded_nodes?

In this case, global_token_metadata_barrier() ensures hard banning by the
means of executing a barrier on each node.

Btw, do we have a scenario where fencing is not enough for banning such
nodes?

**Comment by kbr-scylla** [comment]
> Can hard banning happen lazily as each node learns about excluded_nodes?

IMO yes

**Comment by gleb-cloudius** [comment]
> Can hard banning happen lazily as each node learns about excluded_nodes?

But then how do you know all nodes know not to talk to banned nodes before you move to the next step?

>  In this case, global_token_metadata_barrier() ensures hard banning by the means of executing a barrier on each node.

Which case? Also the barrier may fail as well.

>  Btw, do we have a scenario where fencing is not enough for banning such nodes?

Fencing will only prevent queries from this node to be served by the cluster and only until the node sync up over raft. But we do not want this node to rejoin the cluster any longer and without banning it we cannot guaranty it. Suppose a node 
was down  during a number of topology changes. Now it boots but did not sync with raft yet. It gets queries and replies with errors. This should not happen. A node restart should not cause errors. One way to solve it is to require raft barrier on restart, but this has disadvantages (you can not start a minority cluster any longer). So the idea is to ban the node instead that will prevent it from rejoining the cluster all together. I wanted the banning to be temporary: the node cannot talk to cluster until is syncs with the raft if it was banned. If it wasn't the sync is not required. This is more complicated since it requires to ban only some verbs and still also raft to work, but this will allow us to do topology operation without requiring all nodes to be UP without killing nodes that happen to be down.

So this issue is about banning nodes. You seams to be wanting extensible excluded_nodes list. Are you sure it is actually safe to exclude nodes from barriers without guaranty that they will not be able to communicate with the cluster?

**Comment by tgrabiec** [comment]
On Tue, Jan 30, 2024 at 11:45 AM Gleb Natapov ***@***.***>
wrote:

> Can hard banning happen lazily as each node learns about excluded_nodes?
>
> But then how do you know all nodes know not to talk to banned nodes before
> you move to the next step?
>
global_token_metadata_barrier() ensures that, it waits for all nodes to
learn (and ban) excluded_nodes.



> In this case, global_token_metadata_barrier() ensures hard banning by the
> means of executing a barrier on each node.
>
> Which case? Also the barrier may fail as well.
>

In case we do the lazy hard banning based on excluded_nodes.

If the barrier fails, we don't move to the next step. The cluster either
heals and we can proceed, or the admin needs to exclude more nodes.


> Btw, do we have a scenario where fencing is not enough for banning such
> nodes?
>
> Fencing will only prevent queries from this node to be served by the
> cluster and only until the node sync up over raft. But we do not want this
> node to rejoin the cluster any longer and without banning it we cannot
> guaranty it. Suppose a node
> was down during a number of topology changes. Now it boots but did not
> sync with raft yet. It gets queries and replies with errors. This should
> not happen. A node restart should not cause errors. One way to solve it is
> to require raft barrier on restart, but this has disadvantages (you can not
> start a minority cluster any longer). So the idea is to ban the node
> instead that will prevent it from rejoining the cluster all together. I
> wanted the banning to be temporary: the node cannot talk to cluster until
> is syncs with the raft if it was banned. If it wasn't the sync is not
> required. This is more complicated since it requires to ban only some verbs
> and still also raft to work, but this will allow us to do topology
> operation without requiring all nodes to be UP without killing nodes that
> happen to be down.
>

I think temporary banning cannot work in general, because if
topology operations ignore the node, invariants on that node will not be
maintained. For example, if there is a tablet replica on banned node, we
should clean it up before completing tablet migrations. If migration
ignores the node, it assumes it's safe to not clean it up. That is only
true if the node never comes back. If it comes back, we have data
resurrection.


> So this issue is about banning nodes. You seams to be wanting extensible
> excluded_nodes list. Are you sure it is actually safe to exclude nodes from
> barriers without guaranty that they will not be able to communicate with
> the cluster?
>
It's safe if either (1) node is really down and is not coming up (trust the
admin) or (2) fencing covers all unsafe operations (does it?).

If we're worried about (2) then the solution could be to extend fencing to
more operations, which is also achieved by your "hard banning".

**Comment by kbr-scylla** [comment]
We had a call with Gleb and Tomek and agreed on the following solution.

Suppose the admin does `removenode X --ignore-dead-nodes Y, Z, ...`
At the start of the operation, we will:
- make nodes `X, Y, Z, ...` group 0 non-voters
- add `X, Y, Z, ...` to a global list of excluded nodes

All nodes in the excluded list will also be banned, and tablet load balancer can unblock by ignoring nodes from  this list.

Then `removenode` will proceed as normal.

The solution has simple semantics and is easy to implement. And it simplifies reasoning about other subsystems -- we don't need to worry what happens if a node that we ignored during topology operations comes back later, if it resurrects data or tombstones etc.

The downside is that the `removenode` request may fail / get cancelled, however, the excluded nodes will stay in the list.

I argued that this is not a problem. Our documentation already says that you should only remove nodes / ignore them as dead if they are permanently dead. We just turn this assumption into something that we ensure through banning/exclusion. By executing `removenode X --ignore-dead-nodes Y, Z, ...`, the administrator commits to the decision that nodes `X, Y, Z` will no longer be allowed in the cluster.

If the request fails, the administrator should retry -- if it failed because yet another node was lost permanently, they can extend their ignore list.

Note: Gleb said he'd implement it, but will not enjoy it.
